cranelift/chaos/src/enabled.rs

```diff
@@ -0,0 +1,171 @@
+
+struct ChaosEngineData {
+    /// # Safety
+    ///
+    /// This field must never be moved from, as it is referenced by
+    /// the field `unstructured` for its entire lifetime.
+    ///
+    /// This pattern is the ["self-referential" type](
+    /// https://morestina.net/blog/1868/self-referential-types-for-fun-and-profit)
+    #[allow(dead_code)]
+    data: Vec<u8>,
+    /// We use internal mutability such that a `ChaosEngine` can be passed
+    /// through the call stack without having to be declared as mutable.
+    /// Besides the convenience, the mutation of the internal unstructured
+    /// data should be opaque to users anyway.
+    ///
+    /// # Safety
+    ///
+    /// The lifetime of this is actually not static, but tied to `data`.
+    unstructured: Mutex<Unstructured<'static>>,
+}
+
+impl Debug for ChaosEngineData {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let data_len = self.data.len();
+        let remaining_len = self
+            .unstructured
+            .lock()
+            .expect("poisoned ChaosEngineData mutex")
+            .len();
+        let consumed_len = data_len - remaining_len;
+        f.debug_struct("ChaosEngineData")
+            .field("data", &self.data)
+            .field(
+                "unstructured",
+                &format!(
+                    "(consumed {consumed_len}/{data_len} bytes, \
+                    {remaining_len} remaining)"
+                ),
+            )
+            .finish()
+    }
+}
+
+/// The control plane of chaos mode.
+/// Please see the [crate-level documentation](crate).
+///
+/// **Clone liberally!** The chaos engine is reference counted.
+#[derive(Debug, Clone)]
+pub struct ChaosEngine {
+    data: Arc<ChaosEngineData>,
+    is_todo: bool,
+}
+
+impl ChaosEngine {
+    fn new(data: Vec<u8>, is_todo: bool) -> Self {
+        let unstructured = Unstructured::new(&data);
+        // safety: this is ok because we never move out of the vector
+        let unstructured = Mutex::new(unsafe { std::mem::transmute(unstructured) });
+        Self {
+            data: Arc::new(ChaosEngineData { data, unstructured }),
+            is_todo,
+        }
+    }
+
+    /// This is a zero-sized dummy for use during any builds without the
+    /// feature `chaos` enabled, especially release builds. It has no
+    /// functionality, so the programmer is prevented from using it in any
+    /// way in release builds, which could degrade performance.
+    ///
+    /// This should not be used on code paths that may execute while the
+    /// feature `chaos` is enabled. That would break the assumption that
+    /// [ChaosEngine] is a singleton, responsible for centrally managing
+    /// the pseudo-randomness injected at runtimme.
+    ///
+    /// Use [todo](ChaosEngine::todo) instead, for stubbing out code paths
+    /// you don't expect to be reached (yet) during chaos mode fuzzing.
+    ///
+    /// # Pancis
```

Comment:
```suggestion
    /// # Panics
```
(hi, i was interested in the Unstructured conversation and noticed this)

Reply:
Thanks! fixed it.